src/components/ScatterLayer.tsx

```diff
@@ -0,0 +1,128 @@
+import * as React from 'react'
+import {useRef, useLayoutEffect, useMemo, FunctionComponent} from 'react'
+import {symbol as d3Symbol, SymbolType} from 'd3-shape'
+
+import {Table, Scale, ScatterLayerConfig} from '../types'
+import {PlotEnv} from '../utils/PlotEnv'
+import {clearCanvas} from '../utils/clearCanvas'
+import {GROUP_COL_KEY} from '../constants'
+
+type ScatterData = Array<{
+  xs: number[] | Float64Array
+  ys: number[] | Float64Array
+  fill: string
+  symbol: SymbolType
+}>
+
+const collectScatterData = (
+  table: Table,
+  xColKey: string,
+  yColKey: string,
+  fillScale: Scale<string, string>,
+  symbolScale: Scale<string, SymbolType>
+): ScatterData => {
+  const xCol = table.columns[xColKey].data
+  const yCol = table.columns[yColKey].data
+  const groupCol = table.columns[GROUP_COL_KEY].data as string[]
+  const groupKeys = new Set(groupCol)
+
+  const result = []
+
+  for (let groupKey of groupKeys) {
```

Comment:
There's a minor optimization we can make here that I realized the other day. Instead of mapping over the table once for every group key, we can map over the table once and pick out points for a group key as we see them.

Here's a very similar change made to the line graph: https://github.com/influxdata/vis/commit/acc6ed3b711413e074c37af1bf59895b5d8c4bd4

I've been thinking of this and a lot of the other optimizations in big O terms, which seem particularly relevant for this library, since our performance problem is usually "the data set is too big". If _m_ is the cardinality of the plot and  _n_ is the length of the passed table, the optimization makes this _O(n)_ rather than _O(mn)_. So it's good for high cardinality plots.